pdf_parser/src/pdf_parser/pdf_extractor.py

```diff
@@ -0,0 +1,243 @@
+import nltk
+import fitz  # PyMuPDF
+from pdf2image import convert_from_path
+import pytesseract
+import pdfplumber
+import re
+from nltk.tokenize import sent_tokenize
+import spacy
+import openai
+import os
+import requests
+import json
+from langchain.prompts import ChatPromptTemplate, SystemMessagePromptTemplate, HumanMessagePromptTemplate
+from langchain_community.chat_models import ChatOpenAI
+from langchain_anthropic import ChatAnthropic
+from langchain.base_language import BaseLanguageModel
+
+nltk.download('punkt_tab')
+
+# Load spaCy NLP model
+nlp = spacy.load('en_core_web_sm')
+
+def extract_text_from_pdf(pdf_path):
+    """
+    Extracts text from a PDF file, handling both text-based and scanned PDFs.
+    """
+    text = ""
+
+    try:
+        # Attempt to extract text directly using PyMuPDF
+        with fitz.open(pdf_path) as doc:
+            for page_num in range(len(doc)):
+                page = doc.load_page(page_num)
+                text += page.get_text()
+
+        # If no text is extracted, assume it's a scanned PDF and use OCR
+        if not text.strip():
+            print("No text found. Attempting OCR...")
+            images = convert_from_path(pdf_path, dpi=300)
+            for image in images:
+                text += pytesseract.image_to_string(image)
+
+    except Exception as e:
+        print(f"Error extracting text: {e}")
+
+    return text
+
+def extract_tables_from_pdf(pdf_path):
+    """
+    Extracts tables from a PDF file using pdfplumber.
+    """
+    tables = []
+
+    try:
+        with pdfplumber.open(pdf_path) as pdf:
+            for page in pdf.pages:
+                page_tables = page.extract_tables()
+                if page_tables:
+                    tables.extend(page_tables)
+    except Exception as e:
+        print(f"Error extracting tables: {e}")
+
+    return tables
+
+def clean_and_structure_text(text):
```

Comment:
this should be separate from the PDF extractor

this is for extracting claims from text, we might get the text other ways

pulling text from pdf should be separate from extracting json claims from text please